magma/backend/mlir/hardware_module.py

```diff
@@ -640,6 +703,10 @@ def _visit(value, counter, value_to_index):
             sv.RegOp(results=[wire])
             sv.ReadInOutOp(operands=[wire], results=[result])
 
+        # Track outer_block so array_ref/constants are placed outside the always
+        # comb to reduce code repetition
+        outer_block = get_block_stack().peek()
```

Comment:
Note, this outer_block logic means that some refs/consts are appended after the always block where they are used, MLIR is okay with this (tested with mlir-verilog compile), but I may look into adding some logic to insert these before the block that references them to make the output more readable